File: maestro/maestro/custom/project.py
# Copyright (c) 2024, DAS and Contributors
# License: GNU General Public License v3. See license.txt

import frappe
from frappe.utils import date_diff

def validate_so_and_daily_operation(self, method):
    if self.do_setting.create_by_purchase_order:
        return 
    
    do_project = self.get("custom_reference", { "daily_operation": True })
    if not do_project:
        return
    
    # validasi daily operation harus memiliki sales order yang sama dengan project 
    do_list = frappe.get_list("Daily Operation", 
        {
            "name": ["in", [r.daily_operation for r in do_project]],
            "sales_order": ["!=", self.sales_order]
        }, pluck="name")
    
    if do_list or len(do_project) != (
        date_diff(self.expected_end_date, self.expected_start_date) + 1
    ):
        frappe.throw("Please recreate the daily operation.")

def create_daily_operation(self, method=None):
    if self.do_setting.create_by_purchase_order:
        return 
    
    not_link_do = self.get("custom_reference", { "daily_operation": False })
    if not not_link_do:
        return
    
    # buat daily operation dengan default value sesuai dengan data pada  
    for row in not_link_do:
        do = frappe.new_doc("Daily Operation")
        # field dari doc project
        for field in ["sales_order", "company"]:
            do.set(field, self.get(field))

        # field dari table references
        for field in ["date", "hotel", "lunch", "flight", "currency", "conversion_rate", "grand_total"]:
            do.set(field, row.get(field))

        do.update_base_total()
        do.save()

        row.daily_operation = do.name
        row.base_grand_total = do.base_grand_total
        
    # check data reference pada database untuk d bandingkan
    previous = self.get_doc_before_save()
    if not previous or not previous.get("custom_reference"):
        return
    
    # get data daily operation yang sudah tidak ada pada table untuk d hapus
    removed_do = list(set([x.daily_operation for x in previous.custom_reference]) - set([x.daily_operation for x in self.custom_reference]))    
    if not removed_do:
        return

    for row in removed_do:
        frappe.delete_doc("Daily Operation", row, force=1)

def remove_daily_operation(self, method=None):
    if self.do_setting.create_by_purchase_order:
        return 
    
    link_do = self.get("custom_reference", { "daily_operation": True })
    
    for row in link_do:
        frappe.delete_doc("Daily Operation", row.daily_operation)

File: maestro/maestro/custom/project.js
// Copyright (c) 2024, DAS and contributors
// For license information, please see license.txt

frappe.ui.form.on("Project", {
    refresh: function (frm) {
        frm.set_df_property("custom_reference", "cannot_add_rows", true);
		frm.set_df_property("custom_reference", "cannot_delete_rows", true);

        frm.set_query("hotel", "custom_reference", function (doc, cdt, cdn) {
            // var item = locals[cdt][cdn]

			return {
				filters: {
					item_group: "Hotels",
				},
			};
		});
    },
    custom_create_daily_operation(frm){
        if(!frm.doc.expected_start_date || ! frm.doc.expected_end_date){
            frappe.throw("Please enter the check-in and check-out dates.")
        }

        if((frm.doc.custom_reference || []).length > 0){
            erpnext.project.show_prompt_reset_daily_operation(frm)
        }else{
            erpnext.project.create_daily_operation(frm)
        }
    }
})

erpnext.project = {
    show_prompt_reset_daily_operation : function(frm) {
        frappe.warn('The Daily Operation will be removed.',
            'Are you certain you want to proceed?',
            () => {            
                erpnext.project.create_daily_operation(frm)
            },
            'Continue'
        )
    },
    create_daily_operation: function(frm){
        frappe.call({
            doc: frm.doc,
            method: "create_daily_operation",
            freeze: true,
            callback: function () {
                frm.doc.__unsaved = 1
                frm.get_field("custom_reference").grid.reset_grid()
            },
        })
    }
}